contracts/oracle/CollateralizationOracle.sol

```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "./UniswapOracle.sol";
+import "../refs/CoreRef.sol";
+import "../external/SafeMathCopy.sol";
+import "../pcv/EthUniswapPCVDeposit.sol";
+
+/// @title Collateralization Oracle
+/// @author eswak
+contract CollateralizationOracle is IOracle, CoreRef {
+    using Decimal for Decimal.D256;
+    using SafeMathCopy for uint256;
+
+    /// @notice reference oracles for various PCV currencies
+    IUniswapOracle public ethPriceOracle;
+
+    // @notice reference EthUniswapPCVDeposit
+    EthUniswapPCVDeposit public ethUniswapPCVDeposit;
+
+    /// @notice UniswapOracle constructor
+    /// @param _core Fei Core for reference
+    /// @param _ethPriceOracle Oracle for ETH price
+    /// @param _ethUniswapPCVDeposit reference to the EthUniswapPCVDeposit contract
+    constructor(
+        address _core,
+        address _ethPriceOracle,
+        address payable _ethUniswapPCVDeposit
+    ) public CoreRef(_core) {
+        ethPriceOracle = UniswapOracle(_ethPriceOracle);
+        ethUniswapPCVDeposit = EthUniswapPCVDeposit(_ethUniswapPCVDeposit);
+    }
+
+    /// @notice get the current circulating suply of FEI
+    /// @return number of FEI in circulation
+    function _circulatingFei() internal view returns(uint256) {
+      (uint256 feiInPool, uint256 ethInPool) = ethUniswapPCVDeposit.getReserves();
+      uint256 ethPcv = ethUniswapPCVDeposit.totalValue();
+      uint256 feiPcv = Decimal.ratio(ethPcv, ethInPool).mul(feiInPool).asUint256();
```

Comment:
Using the spot reserves of Uniswap would be vulnerable to flash manipulation. I'd suggest using an ETH-FEI TWAP oracle and then back out manipulation resistant numbers for the reserves by doing `k = feiInPool * ethInPool`
`resistantEthInPool = sqrt(k/ethFeiOraclePrice)` and `resistantFeiInPool = k/resistantEthInPool`

Here is the address for our current ETH-FEI TWAP: 0xB58cbCb1e98c7F3771847cc2D730bF9589b534e4. Its 30 min so would only be more useful for less price sensitive applications

Reply:
done